libs/parallelism/execution/include/hpx/execution/algorithms/ensure_started.hpp

```diff
@@ -0,0 +1,377 @@
+                operation_state_type os;
+
+                struct done_type
+                {
+                };
+                using value_type =
+                    typename hpx::execution::experimental::sender_traits<
+                        S>::template value_types<hpx::tuple, std::variant>;
+                using error_type =
+                    hpx::util::detail::unique_t<hpx::util::detail::prepend_t<
+                        error_types<std::variant>, std::exception_ptr>>;
+                std::variant<std::monostate, done_type, error_type, value_type>
+                    v;
+
+                using continuation_type =
+                    hpx::util::unique_function_nonser<void()>;
+                boost::container::small_vector<continuation_type, 1>
+                    continuations;
+
+            public:
+                template <typename S_,
+                    typename = std::enable_if_t<
+                        !std::is_same<std::decay_t<S_>, shared_state>::value>>
+                shared_state(S_&& s)
+                  : os(hpx::execution::experimental::connect(
+                        std::forward<S_>(s), ensure_started_receiver{this}))
+                {
+                }
+
+                ~shared_state()
+                {
+                    if (!start_called)
+                    {
+                        std::terminate();
+                    }
+
+                    if (--reference_count == 0)
+                    {
+                        delete this;
+                    }
+                }
+
+            private:
+                template <typename R>
+                struct done_error_value_visitor
+                {
+                    std::decay_t<R> r;
+
+                    void operator()(done_type)
+                    {
+                        hpx::execution::experimental::set_done(std::move(r));
+                    }
+
+                    void operator()(error_type const& e)
+                    {
+                        std::visit(error_visitor<R>{std::forward<R>(r)}, e);
+                    }
+
+                    void operator()(value_type const& ts)
+                    {
+                        std::visit(value_visitor<R>{std::forward<R>(r)}, ts);
+                    }
+
+                    void operator()(std::monostate)
+                    {
+                        std::terminate();
+                    }
+                };
+
+                void set_predecessor_done()
+                {
+                    predecessor_done = true;
+
+                    std::unique_lock<hpx::util::spinlock> l{mtx};
+                    if (!continuations.empty())
+                    {
+                        l.unlock();
+                        for (auto const& continuation : continuations)
+                        {
+                            continuation();
+                        }
+
+                        continuations.clear();
+                    }
```

Comment:
```suggestion
                    boost::container::small_vector<continuation_type, 1> conts;
                    {
                        std::unique_lock<hpx::util::spinlock> l{mtx};
                        std::swap(continuations, conts);
                    }
                    for (auto const& continuation : conts)
                    {
                        continuation();
                    }
```

Reply:
What's the motivation behind this suggestion?

Reply:
Your code accesses `continuations` outside of the lock.

Reply:
It does, but no continuations will be added to the vector after this point. The lock around `continuations.empty()` is enough in this case (there's a separate atomic variable which is set to true when one of `set_done/error/value` has been called and that is checked when adding continuations; if one of the `set_` functions have been called the continuation will be called inline). I realize it might not be 100% clear so let me know if this doesn't make sense.

Reply:
Understood. I still think the suggested change would make it more idiomatic without adding overhead.

Reply:
Default constructor + swap compared to a call to `empty`? Probably not a huge difference, but not nothing. In what sense would the former be more idiomatic?

Reply:
What I meant was that it's not obvious to me why calling `.empy()` needs a lock but actually accessing the container doesn't. Perhaps would adding a comment clarify the code. 

As for the slightly larger overhead of my suggestion: I'd prefer having correct over fast code.

Reply:
I fully agree with "I'd prefer having correct over fast code". However, to the best of my knowledge this is fine. Please let me know if you think it actually is racy! Being on the safe side is absolutely better than possibly wrong.

I've extensively commented the sections where the vector is accessed with hopefully useful explanations. The lock around the call to `empty()` was in fact a red herring. Not even that is needed, and I've changed it accordingly. The important thing is synchronizing with the lock so that threads adding to the vector see modifications to the `predecessor_done` flag correctly.

I also think the same change could be made to `future_data` to reduce the critical section there when adding/calling continuations.